File: src/main/state/store.ts
import fs from 'node:fs';
import { readFileOrNull, writeJsonAtomic } from '../../core/fsx/atomic.ts';
import { AppError, errnoOf, fail } from '../../shared/errors.ts';
import { MIGRATIONS, SCHEMA_VERSION, defaults, type Settings } from './schema.ts';
import { sanitize } from './sanitize.ts';

export type LoadStatus =
  /** No settings file yet - a first run. */
  | 'fresh'
  /** Read cleanly at the current schema version. */
  | 'loaded'
  /** Read at an older schema and migrated forward. */
  | 'migrated'
  /** The primary file was unusable; the previous good copy was used instead. */
  | 'recoveredFromBackup'
  /** Both copies were unusable; the wreckage was set aside, not deleted. */
  | 'quarantined';

export interface StoreHealth {
  status: LoadStatus;
  /** Where a corrupt file was moved to, so the UI can point a user at it. */
  quarantinedTo: string | null;
  /** Last write failure. Non-null means the in-memory state is ahead of disk. */
  writeError: { code: string; message: string } | null;
}

interface Parsed {
  settings: Settings;
  migrated: boolean;
}

interface PendingUpdate {
  mutate: (draft: Settings) => unknown;
  resolve: (value: unknown) => void;
  reject: (error: unknown) => void;
}

/**
 * Parse and migrate, or throw. Returning a blank default for anything
 * unexpected is what turns a stray byte into "all your settings are gone", so
 * this function is deliberately loud and its caller decides recovery policy.
 */
function parse(text: string): Parsed {
  let raw: unknown;
  try {
    raw = JSON.parse(text);
  } catch (cause) {
    fail('stateCorrupt', 'settings file is not valid JSON', {
      reason: cause instanceof Error ? cause.message : String(cause)
    });
  }
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    fail('stateCorrupt', 'settings file is not a JSON object');
  }

  let current = raw as Record<string, unknown>;
  // Absent schema field means the original upstream layout, which is version 1.
  const found = typeof current['schema'] === 'number' ? current['schema'] : 1;

  if (found > SCHEMA_VERSION) {
    // A newer build wrote this. Migrating backwards would silently discard
    // whatever it added, so refuse and leave the file untouched instead.
    fail('stateVersionAhead', 'settings were written by a newer version', {
      found,
      supported: SCHEMA_VERSION
    });
  }

  let version = found;
  while (version < SCHEMA_VERSION) {
    const step = MIGRATIONS[version];
    if (!step) fail('stateCorrupt', 'no migration path', { from: version, to: SCHEMA_VERSION });
    current = step(current);
    const next = typeof current['schema'] === 'number' ? current['schema'] : version + 1;
    if (next <= version) fail('stateCorrupt', 'migration did not advance the schema', { version });
    version = next;
  }

  return { settings: sanitize(current), migrated: found !== SCHEMA_VERSION };
}

/**
 * The one owner of the settings file.
 *
 * Reads are synchronous against an in-memory copy. Writes go through update(),
 * which applies mutations under a single writer and coalesces those arriving
 * together into one durable write - so twenty concurrent scan handlers each
 * doing `state.scans[key] = result` cannot lose nineteen of those writes,
 * which is exactly what a read-modify-write per handler does.
 */
export class SettingsStore {
  #settings: Settings;
  #health: StoreHealth;
  #pending: PendingUpdate[] = [];
  #flushing: Promise<void> | null = null;
  // Declared explicitly rather than as a constructor parameter property:
  // Node's strip-only TypeScript mode erases types but never generates code,
  // so `private readonly file: string` in the parameter list will not run.
  readonly #file: string;

  private constructor(file: string, settings: Settings, health: StoreHealth) {
    this.#file = file;
    this.#settings = settings;
    this.#health = health;
  }

  private static backupPath(file: string): string {
    return `${file}.bak`;
  }

  static open(file: string): SettingsStore {
    const primary = readFileOrNull(file);
    if (primary === null) {
      return new SettingsStore(file, defaults(), {
        status: 'fresh',
        quarantinedTo: null,
        writeError: null
      });
    }

    try {
      const { settings, migrated } = parse(primary);
      return new SettingsStore(file, settings, {
        status: migrated ? 'migrated' : 'loaded',
        quarantinedTo: null,
        writeError: null
      });
    } catch (cause) {
      // stateVersionAhead is not corruption - the file is fine and belongs to a
      // newer build. Never quarantine it; let the caller show the mismatch.
      if (cause instanceof AppError && cause.code === 'stateVersionAhead') throw cause;

      const backup = readFileOrNull(SettingsStore.backupPath(file));
      if (backup !== null) {
        try {
          const { settings } = parse(backup);
          return new SettingsStore(file, settings, {
            status: 'recoveredFromBackup',
            quarantinedTo: null,
            writeError: null
          });
        } catch {
          /* the backup is unusable too; fall through to quarantine */
        }
      }

      // Set the wreckage aside under a timestamped name. It may be the only
      // copy of a hand-curated library, so it is never deleted or overwritten.
      const stamp = new Date().toISOString().replace(/[:.]/g, '-');
      const quarantine = `${file}.corrupt-${stamp}`;
      try {
        fs.renameSync(file, quarantine);
      } catch {
        return new SettingsStore(file, defaults(), {
          status: 'quarantined',
          quarantinedTo: null,
          writeError: null
        });
      }
      return new SettingsStore(file, defaults(), {
        status: 'quarantined',
        quarantinedTo: quarantine,
        writeError: null
      });
    }
  }

  /** The current settings. The only way to change them is update(). */
  get(): Readonly<Settings> {
    return this.#settings;
  }

  health(): Readonly<StoreHealth> {
    return this.#health;
  }

  /**
   * Mutate under the write lock and persist. The draft handed to mutate() is a
   * private copy, so a throwing mutator leaves the live settings untouched.
   *
   * Mutations that arrive together are coalesced into one write. A library
   * scan produces one of these per game, and writing the whole settings file
   * once per result means a hundred-game library does a hundred
   * write-flush-replace cycles to save a hundred small facts - which is slow,
   * and on Windows invites the antivirus to hold the file mid-replace. The
   * batch is applied in arrival order and persisted once.
   */
  update<T>(mutate: (draft: Settings) => T): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      this.#pending.push({
        mutate,
        resolve: resolve as (value: unknown) => void,
        reject
      });
      this.#schedule();
    });
  }

  #schedule(): void {
    if (this.#flushing) return;
    this.#flushing = (async () => {
      // Yield once so every caller in this turn joins the same batch.
      await Promise.resolve();
      while (this.#pending.length > 0) {
        await this.#flush(this.#pending.splice(0, this.#pending.length));
      }
      this.#flushing = null;
    })();
  }

  async #flush(batch: PendingUpdate[]): Promise<void> {
    let draft = structuredClone(this.#settings);
    const outcomes: { entry: PendingUpdate; ok: boolean; value?: unknown; error?: unknown }[] = [];

    for (const entry of batch) {
      // Each mutator runs against its own copy, so one that throws part-way
      // through cannot leave the batch holding half of its changes.
      const attempt = structuredClone(draft);
      try {
        const value = entry.mutate(attempt);
        draft = attempt;
        outcomes.push({ entry, ok: true, value });
      } catch (error) {
        outcomes.push({ entry, ok: false, error });
      }
    }

    const applied = outcomes.filter((outcome) => outcome.ok);
    if (applied.length === 0) {
      for (const outcome of outcomes) outcome.entry.reject(outcome.error);
      return;
    }

    draft.schema = SCHEMA_VERSION;
    const next = sanitize(draft);
    try {
      await this.persist(next);
      this.#settings = next;
    } catch (error) {
      // The batch did not reach disk, so everyone whose mutation it carried
      // must hear about it - not just whoever happened to be last.
      for (const outcome of outcomes) outcome.entry.reject(outcome.ok ? error : outcome.error);
      return;
    }
    for (const outcome of outcomes) {
      if (outcome.ok) outcome.entry.resolve(outcome.value);
      else outcome.entry.reject(outcome.error);
    }
  }

  private async persist(next: Settings): Promise<void> {
    try {
      // Keep the last known-good copy before replacing it. This is the file
      // open() falls back to, and it is what makes a torn write survivable.
      try {
        await fs.promises.copyFile(this.#file, SettingsStore.backupPath(this.#file));
      } catch (cause) {
        if (errnoOf(cause) !== 'ENOENT') throw cause;
      }
      await writeJsonAtomic(this.#file, next);
      this.#health = { ...this.#health, writeError: null };
    } catch (cause) {
      // Surface it. Swallowing this is how a read-only or full disk becomes a
      // silent, permanent loss of every setting the user changes afterwards.
      const error = new AppError('stateUnwritable', 'could not save settings', {
        file: this.#file,
        reason: cause instanceof Error ? cause.message : String(cause)
      });
      this.#health = {
        ...this.#health,
        writeError: { code: error.code, message: error.message }
      };
      throw error;
    }
  }

  /** Wait for every queued write to settle. Used on shutdown and in tests. */
  async drain(): Promise<void> {
    // A flush can enqueue the next batch, so keep waiting until it is idle.
    while (this.#flushing) await this.#flushing;
  }

  /** How many writes are queued. Non-zero at shutdown means call drain(). */
  get queued(): number {
    return this.#pending.length;
  }
}

File: src-tauri/crates/core/src/settings/store.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

use serde::Serialize;
use serde_json::{Map, Value};

use crate::error::{fail, Code, Error, Result};
use crate::fsx::atomic::{read_to_string_or_none, write_json_atomic};

use super::sanitize::sanitize;
use super::schema::{migrate, Settings, SCHEMA_VERSION};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum LoadStatus {
    /// No settings file yet - a first run.
    Fresh,
    /// Read cleanly at the current schema version.
    Loaded,
    /// Read at an older schema and migrated forward.
    Migrated,
    /// The primary file was unusable; the previous good copy was used instead.
    RecoveredFromBackup,
    /// Both copies were unusable; the wreckage was set aside, not deleted.
    Quarantined,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WriteError {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Health {
    pub status: LoadStatus,
    /// Where a corrupt file was moved to, so the UI can point a user at it.
    pub quarantined_to: Option<String>,
    /// Last write failure. Some(..) means memory is ahead of disk.
    pub write_error: Option<WriteError>,
}

struct State {
    settings: Settings,
    health: Health,
}

/// The one owner of the settings file.
///
/// Reads are cheap clones of an in-memory copy. Writes take the lock, so
/// concurrent callers each doing `scans.insert(key, result)` cannot lose one
/// another's work - which is exactly what a read-modify-write per handler does.
pub struct SettingsStore {
    file: PathBuf,
    state: Mutex<State>,
}

/// Parse and migrate, or fail loudly.
///
/// Returning a blank default for anything unexpected is what turns a stray
/// byte into "all your settings are gone", so this is deliberately noisy and
/// the caller decides the recovery policy.
fn parse(text: &str) -> Result<(Settings, bool)> {
    let raw: Value = serde_json::from_str(text)
        .map_err(|error| Error::new(Code::StateCorrupt, format!("not valid JSON: {error}")))?;
    let Some(object) = raw.as_object() else {
        return fail(Code::StateCorrupt, "settings file is not a JSON object");
    };

    // An absent `schema` field means the original upstream layout: version 1.
    //
    // Saturating rather than casting: a hand-edited file claiming schema
    // 2^32 + 1 must read as "from the future" and be refused, not wrap around
    // to a version this build thinks it can migrate.
    let found = object
        .get("schema")
        .and_then(Value::as_u64)
        .map_or(1, |value| u32::try_from(value).unwrap_or(u32::MAX));

    if found > SCHEMA_VERSION {
        // A newer build wrote this. Migrating backwards would silently discard
        // whatever it added, so refuse and leave the file untouched.
        return fail(
            Code::StateVersionAhead,
            format!(
                "written by a newer version (schema {found}, this build supports {SCHEMA_VERSION})"
            ),
        );
    }

    let mut current: Map<String, Value> = object.clone();
    let mut version = found;
    while version < SCHEMA_VERSION {
        let Some(stepped) = migrate(version, &current) else {
            return fail(
                Code::StateCorrupt,
                format!("no migration path from schema {version}"),
            );
        };
        let next = stepped
            .get("schema")
            .and_then(Value::as_u64)
            .and_then(|value| u32::try_from(value).ok())
            .unwrap_or_else(|| version.saturating_add(1));
        if next <= version {
            return fail(Code::StateCorrupt, "migration did not advance the schema");
        }
        current = stepped;
        version = next;
    }

    Ok((sanitize(&Value::Object(current)), found != SCHEMA_VERSION))
}

impl SettingsStore {
    fn backup_path(file: &Path) -> PathBuf {
        let mut name = file.as_os_str().to_owned();
        name.push(".bak");
        PathBuf::from(name)
    }

    pub fn open(file: impl Into<PathBuf>) -> Result<Self> {
        let file = file.into();

        let primary = match read_to_string_or_none(&file) {
            Ok(text) => text,
            // Unreadable-but-present is a real failure, not an absence, and it
            // is handled below like any other unusable primary.
            Err(_) => Some(String::new()),
        };

        let Some(primary) = primary else {
            return Ok(Self::with(
                file,
                Settings::default(),
                Health {
                    status: LoadStatus::Fresh,
                    quarantined_to: None,
                    write_error: None,
                },
            ));
        };

        match parse(&primary) {
            Ok((settings, migrated)) => Ok(Self::with(
                file,
                settings,
                Health {
                    status: if migrated {
                        LoadStatus::Migrated
                    } else {
                        LoadStatus::Loaded
                    },
                    quarantined_to: None,
                    write_error: None,
                },
            )),
            Err(error) if error.code == Code::StateVersionAhead => {
                // Not corruption: the file is fine and belongs to a newer
                // build. Never quarantine it; let the caller show the mismatch.
                Err(error)
            }
            Err(_) => {
                if let Ok(Some(backup)) = read_to_string_or_none(&Self::backup_path(&file)) {
                    if let Ok((settings, _)) = parse(&backup) {
                        return Ok(Self::with(
                            file,
                            settings,
                            Health {
                                status: LoadStatus::RecoveredFromBackup,
                                quarantined_to: None,
                                write_error: None,
                            },
                        ));
                    }
                }

                // Set the wreckage aside under a timestamped name. It may be
                // the only copy of a hand-curated library, so it is never
                // deleted or overwritten.
                let quarantine = Self::quarantine_path(&file);
                let moved = fs::rename(&file, &quarantine).is_ok();
                Ok(Self::with(
                    file,
                    Settings::default(),
                    Health {
                        status: LoadStatus::Quarantined,
                        quarantined_to: moved.then(|| quarantine.to_string_lossy().into_owned()),
                        write_error: None,
                    },
                ))
            }
        }
    }

    fn quarantine_path(file: &Path) -> PathBuf {
        let stamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);
        let mut name = file.as_os_str().to_owned();
        name.push(format!(".corrupt-{stamp}"));
        PathBuf::from(name)
    }

    fn with(file: PathBuf, settings: Settings, health: Health) -> Self {
        Self {
            file,
            state: Mutex::new(State { settings, health }),
        }
    }

    fn lock(&self) -> std::sync::MutexGuard<'_, State> {
        // A poisoned lock means a previous holder panicked mid-update. The
        // settings in there are a sanitised value either way, so recovering is
        // better than propagating a panic into every later call.
        self.state
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
    }

    pub fn get(&self) -> Settings {
        self.lock().settings.clone()
    }

    pub fn health(&self) -> Health {
        self.lock().health.clone()
    }

    /// Mutate under the lock and persist.
    ///
    /// The draft is a private copy, so a mutator that leaves it inconsistent
    /// cannot be committed, and the live settings are replaced only once the
    /// write has reached disk.
    pub fn update<T>(&self, mutate: impl FnOnce(&mut Settings) -> T) -> Result<T> {
        let mut state = self.lock();
        let mut draft = state.settings.clone();
        let outcome = mutate(&mut draft);
        draft.schema = SCHEMA_VERSION;
        let next = sanitize(&serde_json::to_value(&draft).unwrap_or(Value::Null));

        match self.persist(&next) {
            Ok(()) => {
                state.settings = next;
                state.health.write_error = None;
                Ok(outcome)
            }
            Err(error) => {
                // Surface it. Swallowing this is how a read-only or full disk
                // becomes a silent, permanent loss of everything the user
                // changes afterwards.
                state.health.write_error = Some(WriteError {
                    code: error.code.as_str().to_owned(),
                    message: error.detail.clone(),
                });
                Err(error)
            }
        }
    }

    fn persist(&self, next: &Settings) -> Result<()> {
        // Keep the last known-good copy before replacing it. This is the file
        // `open` falls back to, and it is what makes a torn write survivable.
        if self.file.exists() {
            let _ = fs::copy(&self.file, Self::backup_path(&self.file));
        }
        write_json_atomic(&self.file, next)
    }
}
